### PR TITLE
MindtPy log update

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:
This PR improves the log of MindtPy and uniforms the docstring using Google's Python style.

The log of MindtPy now looks like the following.

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
rNLP is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822          -inf           inf      inf%      0.11
         1              MILP           11.7796       11.7796           inf      nan%      0.21
*        1         Fixed NLP           113.389       11.7796       113.389    89.61%      0.27
         2              MILP           53.5457       53.5457       113.389    52.78%      0.31
*        2         Fixed NLP           68.0097       53.5457       68.0097    21.27%      0.39
         3              MILP           65.3819       65.3819       68.0097     3.86%      0.42
Cycling happens after 3 main iterations. The same combination is obtained in iteration 2 This issue happens when the NLP subproblem violates constraint qualification. Convergence to optimal solution is not guaranteed.
Final bound values: LB: 65.38192069756268  UB: 68.00974116542103
 =============================================================================================

```

- ECP
```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
max_binary is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -   Max binary MILP                 5          -inf           inf      inf%      0.04
         1              MILP             -1973         -1973           inf      nan%      0.08
         2              MILP          -1722.18      -1722.18           inf      nan%      0.11
         3              MILP          -1482.46      -1482.46           inf      nan%      0.15
         4              MILP          -1246.61      -1246.61           inf      nan%      0.19
         5              MILP          -1010.35      -1010.35           inf      nan%      0.23
         6              MILP          -780.858      -780.858           inf      nan%      0.27
         7              MILP          -562.079      -562.079           inf      nan%      0.32
         8              MILP          -364.771      -364.771           inf      nan%      0.37
         9              MILP          -225.915      -225.915           inf      nan%      0.42
        10              MILP          -97.9353      -97.9353           inf      nan%      0.46
        11              MILP           3.50341       3.50341           inf      nan%      0.50
        12              MILP           55.8803       55.8803           inf      nan%      0.54
        13              MILP           66.3521       66.3521           inf      nan%      0.58
        14              MILP           67.9592       67.9592           inf      nan%      0.63
        15              MILP           67.9858       67.9858           inf      nan%      0.67
        16              MILP           67.9989       67.9989           inf      nan%      0.72
        17              MILP           68.0054       68.0054           inf      nan%      0.75
        18              MILP           68.0088       68.0088           inf      nan%      0.79
MindtPy-ECP exiting on nonlinear constraints satisfaction. LB: 68.0087549938363 UB: 68.0087549938363

 =============================================================================================

```

- FP standalone 

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
Moving objective to constraint set.
FP is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822          -inf           inf      inf%      0.10
         1            FP-MIP            1.6602          -inf           inf      inf%      0.20
         1            FP-NLP       8.65327e-08          -inf           inf      inf%      0.26
*        1         Fixed NLP           113.389          -inf       113.389      inf%      0.33
         2            FP-MIP                 1          -inf       113.389      inf%      0.36
         2            FP-NLP       6.29988e-08          -inf       113.389      inf%      0.43
*        2         Fixed NLP           104.979          -inf       104.979      inf%      0.51
         3            FP-MIP                 2          -inf       104.979      inf%      0.55
         3            FP-NLP       4.31642e-08          -inf       104.979      inf%      0.62
*        3         Fixed NLP           77.1043          -inf       77.1043      inf%      0.69
         4            FP-MIP                 2          -inf       77.1043      inf%      0.73
         4            FP-NLP        0.00422666          -inf       77.1043      inf%      0.80
         5            FP-MIP           1.10851          -inf       77.1043      inf%      0.85
         5            FP-NLP         0.0063319          -inf       77.1043      inf%      0.92
         6            FP-MIP           2.05136          -inf       77.1043      inf%      0.97
         6            FP-NLP        0.00245847          -inf       77.1043      inf%      1.04
         7            FP-MIP           1.95506          -inf       77.1043      inf%      1.08
         7            FP-NLP       4.10246e-08          -inf       77.1043      inf%      1.15
*        7         Fixed NLP            73.278          -inf        73.278      inf%      1.24
         8            FP-MIP                 1          -inf        73.278      inf%      1.27
         8            FP-NLP        4.3704e-08          -inf        73.278      inf%      1.34
*        8         Fixed NLP           68.0097          -inf       68.0097      inf%      1.43
FP-MIP infeasible
 =============================================================================================

```

- FP + OA

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
Moving objective to constraint set.
FP is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822          -inf           inf      inf%      0.07
         1            FP-MIP            1.6602          -inf           inf      inf%      0.10
         1            FP-NLP       8.65327e-08          -inf           inf      inf%      0.17
*        1         Fixed NLP           113.389          -inf       113.389      inf%      0.25
         2            FP-MIP                 2          -inf       113.389      inf%      0.30
         2            FP-NLP       8.12177e-08          -inf       113.389      inf%      0.38
*        2         Fixed NLP           104.294          -inf       104.294      inf%      0.47
         3            FP-MIP                 1          -inf       104.294      inf%      0.53
         3            FP-NLP       6.74346e-08          -inf       104.294      inf%      0.59
*        3         Fixed NLP           95.8848          -inf       95.8848      inf%      0.67
         4            FP-MIP                 2          -inf       95.8848      inf%      0.75
         4            FP-NLP       4.33075e-08          -inf       95.8848      inf%      0.82
*        4         Fixed NLP           68.0097          -inf       68.0097      inf%      0.91
FP-MIP infeasible
         1              MILP           69.9326       69.9326       68.0097    -2.83%      1.03
MindtPy exiting on bound convergence. LB: 69.93264533571386 + (tol 0.0001) >= UB: 68.00973983066932

Solve the main problem without the last no_good cut to fix the bound.zero_tolerance is set to 1E-4
         5         Fixed NLP            73.278       69.9326       68.0097    -2.83%      1.10
Fixed bound values: LB: 73.2779865512031  UB: 68.00973983066932
 =============================================================================================

```

- GOA

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
rNLP is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822          -inf           inf      inf%      0.11
         1              MILP            22.197        22.197           inf      nan%      0.23
*        1         Fixed NLP           104.294        22.197       104.294    78.72%      0.29
         2              MILP           60.3563       60.3563       104.294    42.13%      0.34
*        2         Fixed NLP           77.1043       60.3563       77.1043    21.72%      0.41
         3              MILP           68.0097       68.0097       77.1043    11.80%      0.45
*        3         Fixed NLP           68.0097       68.0097       68.0097    -0.00%      0.53
MindtPy exiting on bound convergence. LB: 68.00974404989009 + (tol 0.0001) >= UB: 68.00974203706781

Solve the main problem without the last no_good cut to fix the bound.zero_tolerance is set to 1E-4
Fixed bound values: LB: 68.00974404989009  UB: 68.00974203706781
 =============================================================================================

```

- ROA(level L1)

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
Moving objective to constraint set.
rNLP is the initial strategy being used.

 =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822          -inf           inf      inf%      0.12
         1              MILP           11.7796       11.7796           inf      nan%      0.24
*        1         Fixed NLP           113.389       11.7796       113.389    89.61%      0.32
         2              MILP           53.5457       53.5457       113.389    52.78%      0.38
         2          Reg MILP           9.78602       53.5457       113.389    52.78%      0.43
*        2         Fixed NLP           77.1043       53.5457       77.1043    30.55%      0.51
         3              MILP           54.8613       54.8613       77.1043    28.85%      0.55
         3          Reg MILP           12.0984       54.8613       77.1043    28.85%      0.61
*        3         Fixed NLP           68.0097       54.8613       68.0097    19.33%      0.68
         4              MILP           65.3819       65.3819       68.0097     3.86%      0.74
         4          Reg MILP           1.31391       65.3819       68.0097     3.86%      0.79
Cycling happens after 4 main iterations. The same combination is obtained in iteration 3 This issue happens when the NLP subproblem violates constraint qualification. Convergence to optimal solution is not guaranteed.
Final bound values: LB: 65.38192069755121  UB: 68.00974116772524
 =============================================================================================

```
 
- LP/NLP

```
---------------------------------------------------------------------------------------------
              Mixed-Integer Nonlinear Decomposition Toolbox in Pyomo (MindtPy)               
---------------------------------------------------------------------------------------------
For more information, please visit https://pyomo.readthedocs.io/en/stable/contributed_packages/mindtpy.html
Original model has 31 constraints (5 nonlinear) and 0 disjunctions, with 32 variables, of which 8 are binary, 0 are integer, and 24 are continuous.
rNLP is the initial strategy being used.

  =============================================================================================
 Iteration | Subproblem Type | Objective Value | Lower Bound | Upper Bound |   Gap   | Time(s)

         -       Relaxed NLP           15.0822       15.0822           inf      nan%      0.11
         1           restrLP           15.0822       15.0822           inf      nan%      0.16
*        1         Fixed NLP           113.389       15.0822       113.389    86.70%      0.28
         1           restrLP           113.356       15.0822       113.389    86.70%      0.29
         1           restrLP           22.7796       15.0822       113.389    86.70%      0.30
*        2         Fixed NLP           85.5139       15.0822       85.5139    82.36%      0.44
         1           restrLP           85.4813       15.0822       85.5139    82.36%      0.44
         1           restrLP           74.4813       17.3549       85.5139    79.71%      0.45
         3         Fixed NLP           113.389       17.3549       85.5139    79.71%      0.54
         1           restrLP           53.5457       53.5457       85.5139    37.38%      0.55
*        4         Fixed NLP           68.0097       53.5457       68.0097    21.27%      0.69
         1           restrLP           65.3819       57.4194       68.0097    15.57%      0.69
         5         Fixed NLP           68.0097       57.4194       68.0097    15.57%      0.82
MindtPy unable to converge bounds after 1 main iterations.
Final bound values: LB: 65.38192069755974  UB: 68.00974116542099
 =============================================================================================

```

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/mindtpy/MindtPy.py [hunk 1]
```diff
@@ -35,14 +35,14 @@
...
-from pyomo.contrib.mindtpy.util import model_is_valid, setup_solve_data
+from pyomo.contrib.mindtpy.util import model_is_valid, set_up_solve_data
 from pyomo.core import (Block, ConstraintList, NonNegativeReals,
                         Set, Suffix, Var, VarList, TransformationFactory, Objective, RangeSet)
 from pyomo.opt import SolverFactory
 from pyomo.contrib.mindtpy.config_options import _get_MindtPy_config, check_config
 
 logger = logging.getLogger('pyomo.contrib.mindtpy')
-
+logging.basicConfig(format='%(message)s')
```
Comment (jsiirola) [inline pyomo/contrib/mindtpy/MindtPy.py:45]: This is almost certainly not a good idea: importing Mindtpy now reconfigures all logging across Python.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/MindtPy.py:45]: Actually, I only want to change the logging format in MindtPy. Do you have suggestions for this?

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/MindtPy.py:45]: Btw, do we need `logger = logging.getLogger('pyomo.contrib.mindtpy')` in each file of MindtPy? 
I remember GDPOpt previously had similar codes. But they are removed now. 
I am not very familiar with the `logging` package. Any comments or suggestions are welcomed. Thanks.

**Comment by ZedongPeng** [comment]
Hi @jsiirola . I tried to avoid defining the global logging config. But I am not sure it's the correct way. It can absolutely be improved. Let me know if you have any suggestions. Thanks.

**Comment by mrmundt** [comment]
@ZedongPeng - We have resolved the issue that was causing the OSX failures. If you merge in changes on `main`, you should be fine.

**Comment by ZedongPeng** [comment]
> @ZedongPeng - We have resolved the issue that was causing the OSX failures. If you merge in changes on `main`, you should be fine.

Thanks for your information. @mrmundt

**Review by jsiirola — APPROVED** [review]
Some minor comments, but nothing I feel should prevent merging.

File: pyomo/contrib/mindtpy/iterate.py [hunk 2]
```diff
@@ -299,6 +286,7 @@ def algorithm_should_terminate(solve_data, config, check_cycling):
         return True
 
     # Check if algorithm is stalling
+    # BUG TODO: here only applies to minimization problems.
```
[the remark this answers is not in the document]
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/iterate.py]: Sure. I just fixed this bug.

File: pyomo/contrib/mindtpy/config_options.py [hunk 3]
```diff
@@ -96,6 +101,12 @@ def _get_MindtPy_config():
         description='The logger object or name to use for reporting.',
         domain=a_logger
     ))
+    CONFIG.declare('logging_level', ConfigValue(
+        default=logging.INFO,
+        domain=In([logging.NOTSET, logging.DEBUG, logging.INFO,
+                   logging.WARNING, logging.ERROR, logging.CRITICAL]),
```
[the remark this answers is not in the document]
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/config_options.py:109]: Thanks for your information. Should I relax the current domain to `[0, 1, .., 50]` or just `NonNegativeInt`?

[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/contrib/mindtpy/config_options.py:109]: I *think* that logging only requires `int`, although I have never tried passing in a negative number before.  I think `NonNegativeInt` is probably the right path for now (at least until someone complains).

File: pyomo/contrib/mindtpy/util.py [hunk 4]
```diff
@@ -24,38 +24,32 @@
...
     pass
 
 
 def model_is_valid(solve_data, config):
-    """
-    Determines whether the model is solveable by MindtPy.
-
-    This function returns True if the given model is solveable by MindtPy (and performs some preprocessing such
-    as moving the objective to the constraints).
+    """Determines whether the model is solveable by MindtPy.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:40]: Typo: solvable

File: pyomo/contrib/mindtpy/util.py [hunk 5]
```diff
@@ -24,38 +24,32 @@
...
+    solve_data : MindtPySolveData
+        Data container that holds solve-instance data.
+    config : ConfigBlock
+        The specific configurations for MindtPy.
 
     Returns
     -------
-    Boolean value (True if model is solveable in MindtPy else False)
+    bool
+        True if model is solveable in MindtPy, False otherwise.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:52]: Typo: solvable

File: pyomo/contrib/mindtpy/util.py [hunk 6]
```diff
@@ -193,18 +184,23 @@ def var_bound_add(solve_data, config):
...
-    setpoint_model: Pyomo model
-        the model that provides the base point for us to calculate the distance
-    discrete_only: Bool
-        only optimize on distance between the discrete variables
+    model : Pyomo model
+        The model that needs new objective function.
+    setpoint_model : Pyomo model
+        The model that provides the base point for us to calculate the distance.
+    discrete_only : bool, optional
+        Whether only optimize on distance between the discrete variables, by default False.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:198]: Typo: "whether to"

File: pyomo/contrib/mindtpy/util.py [hunk 7]
```diff
@@ -224,18 +220,23 @@ def generate_norm2sq_objective_function(model, setpoint_model, discrete_only=Fal
...
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:234]: Typo: "whether to"

File: pyomo/contrib/mindtpy/util.py [hunk 8]
```diff
@@ -300,12 +306,25 @@ def generate_norm_inf_objective_function(model, setpoint_model, discrete_only=Fa
...
+    model : Pyomo model
+        The model that needs new objective function.
+    setpoint_model : Pyomo model
+        The model that provides the base point for us to calculate the distance.
+    config : ConfigBlock
+        The specific configurations for MindtPy.
+    solve_data : MindtPySolveData
+        Data container that holds solve-instance data.
+    discrete_only : bool, optional
+        Whether only optimize on distance between the discrete variables, by default False.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:322]: Typo: "whether to"

File: pyomo/contrib/mindtpy/util.py [hunk 9]
```diff
@@ -374,22 +393,23 @@ def generate_lag_objective_function(model, setpoint_model, config, solve_data, d
...
-    discrete_only: Bool
-        only optimize on distance between the discrete variables
+    model : Pyomo model
+        The model that needs the norm constraint.
+    setpoint_model : Pyomo model
+        The model that provides the base point for us to calculate the distance.
+    config : ConfigBlock
+        The specific configurations for MindtPy.
+    discrete_only : bool, optional
+        Whether only optimize on distance between the discrete variables, by default True.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:411]: Typo: "whether to"

File: pyomo/contrib/mindtpy/util.py [hunk 10]
```diff
@@ -619,24 +663,24 @@ def setup_solve_data(model, config):
...
+    Parameters
+    ----------
+    from_list : list
+        The variables that provides the values to copy from.
+    to_list : list
+        The variables that need to set value.
+    config : ConfigBlock
+        The specific configurations for MindtPy.
+    solver_model : solver model
+        The solver model derive from pyomo model.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:677]: typo: derived

File: pyomo/contrib/mindtpy/util.py [hunk 11]
```diff
@@ -619,24 +663,24 @@ def setup_solve_data(model, config):
...
+    var_map : dict
+        The map of pyomo variables to solver variables.
+    solution_name : int or str
+        The name of the solution in the solution pool.
+    ignore_integrality : bool, optional
+        Whether ignore the integrality of integer variables, by default False.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:683]: typo: whether to

File: pyomo/contrib/mindtpy/util.py [hunk 12]
```diff
@@ -672,9 +716,125 @@ def copy_var_list_values_from_solution_pool(from_list, to_list, config, solver_m
...
+    """
+    solve_data.abs_gap = solve_data.UB - solve_data.LB
+    solve_data.rel_gap = (solve_data.UB - solve_data.LB)/(abs(
+        solve_data.UB if solve_data.objective_sense == minimize else solve_data.LB) + 1E-10)
+
+
+def update_dual_bound(solve_data, bound_value):
+    """Update the dual bound.
+
+    Call after solve relaxed problem, including relaxed NLP and MIP master problem.
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:754]: typo: solving

File: pyomo/contrib/mindtpy/util.py [hunk 13]
```diff
@@ -672,9 +716,125 @@ def copy_var_list_values_from_solution_pool(from_list, to_list, config, solver_m
...
+        solve_data.LB_progress.append(solve_data.LB)
+    else:
+        solve_data.UB = min(bound_value, solve_data.UB)
+        solve_data.bound_improved = solve_data.UB < solve_data.UB_progress[-1]
+        solve_data.UB_progress.append(solve_data.UB)
+    if solve_data.bound_improved:
+        update_gap(solve_data)
+
+
+def uptade_suboptimal_dual_bound(solve_data, results):
```
Comment (emma58) [inline pyomo/contrib/mindtpy/util.py:778]: typo: `update_suboptimal_dual_bound`

**Review by emma58 — APPROVED** [review]
This looks good. I left a lot of comments, but they're all grammatical or (probably naive) questions.

**Pull request merged**